app/components/model-renderer.jsx

```diff
@@ -0,0 +1,127 @@
+import React from 'react';
+import Model from './modelling';
+
+// function to check whether webgl is available
+function webGLCompatibilityTest() {
+  try {
+    return !!window.WebGLRenderingContext && !!document.createElement('canvas').getContext('experimental-webgl');
+  } catch (e) {
+    return false;
+  }
+}
+
+class ModelRenderer extends React.Component {
+  constructor(props) {
+    super(props);
+    this.getDifference = this.getDifference.bind(this);
+    this.setTexture = this.setTexture.bind(this);
+
+    this.imgUrl = Object.values(this.props.subject.locations[0])[0];
+    if (this.props.subject.locations.length < 2) {
+      const im0 = this.props.subject.locations[0];
+      this.props.subject.locations[1] = Object.assign({}, im0);
+    }
+  }
+  componentDidMount() {
+    if (!webGLCompatibilityTest()) {
+      // TODO: is there a fallback option?
+      /* eslint-disable no-console */
+      console.error('WebGL is not available, aborting');
+      /* eslint-enable no-console */
+      return;
+    }
+    // component has mounted, initialise the regl canvases
+    this.model = new Model(this.canvas, this.props.subject.metadata);
+    // if the image has loaded
+    if (this.imageLoaded) {
+      // provide the differece caluculator with the image
+      this.model.setBaseTexture(this.im);
+      this.model.update(this.props.classification.annotations);
+    } else {
+      // tell the setTexture function that it should update the difference calc
+      this.imageLoaded = true;
+    }
+  }
+  componentWillReceiveProps() {
+    // component has updated with new props, set the new render functions
+    // check if a model has been assigned
+    if (this.model !== null) {
+      // if so, update the render function list to the new annotation
+      window.requestAnimationFrame(this.getDifference);
+    }
+  }
+
+  // don't want to re-render whenever props update
+  shouldComponentUpdate() {
+    if (this.imgUrl !== Object.values(this.props.subject.locations[0])[0]) {
```

Comment:
If `imgUrl` is in state you can check against `nextProps` and `nextState` directly instead of re-fetching the URL from the subject.